sale_order_type/models/sale_order.py

```diff
@@ -53,21 +53,27 @@ def onchange_type_id(self):
 
     @api.model
     def create(self, vals):
-        if vals.get('name', '/') == '/'and vals.get('type_id'):
+        if vals.get('name', '/') == '/' and vals.get('type_id'):
             sale_type = self.env['sale.order.type'].browse(vals['type_id'])
             if sale_type.sequence_id:
                 vals['name'] = sale_type.sequence_id.next_by_id()
         return super(SaleOrder, self).create(vals)
 
     @api.multi
     def write(self, vals):
-        res = super(SaleOrder, self).write(vals)
-        if vals.get("type_id") and (self.state == "draft" or self.state == "sent"):
-            for order in self:
-                sale_type = self.env["sale.order.type"].browse(vals["type_id"])
-                if sale_type.sequence_id:
-                    order.name = sale_type.sequence_id.next_by_id()
-        return res
+        if vals.get("type_id"):
+            sale_type = self.env["sale.order.type"].browse(vals["type_id"])
+            if sale_type.sequence_id:
+                for record in self:
+                    if (
+                        record.state in {"draft", "sent"}
+                        and record.type_id != sale_type
+                    ):
+                        new_vals = vals.copy()
+                        new_vals["name"] = sale_type.sequence_id.next_by_id()
+                        super(SaleOrder, record).write(new_vals)
+                    else:
+                        super(SaleOrder, record).write(vals)
```

Comment:
Missing else part calling super as is. Also missing `return True for this part.

Reply:
Aarrrgghhh sorry. `return True` should be to all function, isn't it?

Reply:
Well, it can be an option, the other is:
```python
if ...:
    return True
else:
   return super().write(vals)
```

Reply:
done, thx again